src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs

```diff
@@ -0,0 +1,91 @@
+using Azure.Core;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Text;
+using System.Threading.Tasks;
+using System;
+using BaGetter.Core;
+using System.Linq;
+
+namespace BaGetter.Web.Authentication;
+
+public class NugetBasicAuthenticationHandler : AuthenticationHandler<AuthenticationSchemeOptions>
+{
+    private readonly IOptions<BaGetterOptions> bagetterOptions;
+
+    public NugetBasicAuthenticationHandler(
+        IOptionsMonitor<AuthenticationSchemeOptions> options,
+        ILoggerFactory logger,
+        UrlEncoder encoder,
+        IOptions<BaGetterOptions> bagetterOptions)
+        : base(options, logger, encoder)
+    {
+        this.bagetterOptions = bagetterOptions;
+    }
+
+    protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        if (bagetterOptions.Value.Authentication is null || bagetterOptions.Value.Authentication.Length==0 || bagetterOptions.Value.Authentication.All(a=> string.IsNullOrWhiteSpace(a.Username) && string.IsNullOrWhiteSpace(a.Password)))
+        {
+            var claims = new[]
+            {
+                new Claim(ClaimTypes.Anonymous, string.Empty),
+            };
+            var identity = new ClaimsIdentity(claims, Scheme.Name);
+            var principal = new ClaimsPrincipal(identity);
+
+            var ticket = new AuthenticationTicket(principal, Scheme.Name);
+
+            return Task.FromResult(AuthenticateResult.Success(ticket));
+        }
+        else
+        {
+            if (!Request.Headers.TryGetValue("Authorization", out var auth))
+                return Task.FromResult(AuthenticateResult.NoResult());
+
+            string username = null;
+            string password = null;
+            try
+            {
+                var authHeader = AuthenticationHeaderValue.Parse(auth);
+                var credentialBytes = Convert.FromBase64String(authHeader.Parameter);
+                var credentials = Encoding.UTF8.GetString(credentialBytes).Split([':'], 2);
+                username = credentials[0];
+                password = credentials[1];
```

Comment:
Wouldn't this fail the authentication when a package is pushed using the API key?

Reply:
No, I have tested that. I did decorate only controllers that handle downloads and browsing. Uploading is not affected. You can even add -username and -password to nuget push.

Reply:
Ok I see. Can this be used for the razor pages too?

Reply:
Not directly like that. It would need to be implemented differently